streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala

```diff
@@ -581,6 +579,9 @@ class StreamingContext private[streaming] (
       case INITIALIZED =>
         startSite.set(DStream.getCreationSite())
         sparkContext.setCallSite(startSite.get)
+        // Registering Streaming Metrics at the start of the StreamingContext
+        assert(env.metricsSystem != null)
+        env.metricsSystem.registerSource(streamingSource)
         StreamingContext.ACTIVATION_LOCK.synchronized {
```

Comment:
This is still a little flawed. there is a case where two StreamingContext can be created, but only one should successfully start. So if one is already started (source registered), the second streaming context.start() should fail. But currently it will fail due to `StreamingContext.assertNoOtherContextIsActive()` but it will attempt to register the source nonetheless. This is bad. So the registering of the source should be done only after that line.

I think it should just before `uiTab.foreach(_.attachI())` line 601

Apologies for changing this so many times.